WireAPI/Sources/WireAPI/Models/Backend/ServerTrustValidator.swift

```diff
@@ -0,0 +1,112 @@
+
+    private let pinnedKeys: [PinnedKey]
+
+    init(pinnedKeys: [PinnedKey]) {
+        self.pinnedKeys = pinnedKeys
+    }
+
+    /// Verifies the server `trust` for the given `host`.
+    ///
+    /// - Parameter trust: The `SecTrust` of the server.
+    /// - Parameter host: The host of the server.
+    /// - Throws: An error if server certificate should not be trusted.
+    /// - Note: If no pinned keys are found for the `host`, the server certificate is trusted.
+
+    func validate(trust: SecTrust, host: String) async throws {
+        let matches = pinnedKeys.filter { $0.matches(host: host) }
+
+        // If no keys are pinned for `host`, we trust the server certificate
+        guard !matches.isEmpty else { return }
+
+        try await Self.verifyServerCertificateTrusted(trust)
+
+        let publicKey = try Self.publicKeyAssociatedWithServerTrust(trust)
+
+        let publicKeys = try matches.map { try Self.certificateKey(for: $0) }
+        guard publicKeys.contains(publicKey) else {
+            throw Error.noMatchingPublicKey
+        }
+    }
+
+    // MARK: - Private
+
+    private static func verifyServerCertificateTrusted(_ serverTrust: SecTrust) async throws {
+        try await withCheckedThrowingContinuation { continuation in
+            // `SecTrustEvaluateAsyncWithError` requires the completion queue to be the same as the queue on which it
+            // is called.
+            let queue = DispatchQueue.global()
+            queue.async {
+                SecTrustEvaluateAsyncWithError(serverTrust, queue) { _, success, error in
+                    if success {
+                        continuation.resume()
+                    } else {
+                        continuation.resume(throwing: Error.evaluatingServerTrustFailed(error))
+                    }
+                }
+            }
+        }
+    }
+
+    /// Returns the public key of the leaf certificate associated with `serverTrust`.
+    ///
+    /// - Parameter serverTrust: SecTrust of server
+    /// - Returns: public key from `serverTrust`
+
+    private static func publicKeyAssociatedWithServerTrust(_ serverTrust: SecTrust) throws -> SecKey {
+        let certificates = SecTrustCopyCertificateChain(serverTrust) ?? [] as CFArray
+        let policy = SecPolicyCreateBasicX509()
+        var secTrust: SecTrust?
+
+        guard SecTrustCreateWithCertificates(certificates, policy, &secTrust) == noErr,
+              let trust = secTrust,
+              let result = SecTrustCopyKey(trust)
+        else {
+            throw Error.noPublicKeyOnServerTrust
+        }
+
+        return result
+    }
+
+    private static func certificateKey(for value: PinnedKey) throws -> SecKey {
+        guard let certificate = SecCertificateCreateWithData(nil, value.key as CFData) else {
+            throw Error.cannotLoadCertificate
+        }
+
+        guard let publicKey = SecCertificateCopyKey(certificate) else {
+            throw Error.cannotRetrievePublicKey
+        }
```

Comment:
I guess a difference with the existing implementation is that is currently done once, but here (I assume) it's done for every challenge. Or is it that there is only a single challenge when first contacting the server? So it practice it's not called often?

Reply:
Correct it would be done for every challenge. But in practice (limited testing) I've only noticed a single challenge when first contacting the server. The alternatives I can think of are storing on the `PinnedKey` (not ideal due to breaking auto Sendable conformance), or caching it somewhere. I'm not sure it is worth it unless we notice a problem in practice.

Reply:
I've thought about it a bit more. I'm actually going to move `SecKey` into `PinnedKey` and mark as @preconcurrency. This has the following advantages:

1. Fixes possible issues you mention
2. Allows for us to crash early on initialization if that is what we want - I think this is probably a good idea for the bundled data.
3. Potentially allows us to check this at compile time once we start using macros